tasmota/tasmota_xnrg_energy/xnrg_25_mk_sky_blu.ino

```diff
@@ -0,0 +1,783 @@
+  const char MKSB_HTTP_SNS_str_m_int[]          PROGMEM = "{s}%s " "{m}" "%d{e}";
+
+/*
+const char * mksb_status_bits[] = {
+                                    // mode
+  "",                               // b0 
+  "",                               // b1
+  D_CHARGE,                         // b2 MPPT Mode Active
+  "",                               // b3
+
+  "",                               // b4
+  "",                               // b5
+  "",                               // b6
+  "",                               // b7
+
+                                            // error
+  D_BATTERY " " D_VOLTAGE " " D_STR_LOW,    // b0 Battery undervoltage (< register 5 / 6)
+  D_BATTERY " " D_VOLTAGE " " D_STR_HIGH,   // b1 Battery overvoltage (> register 1)
+  "",                                       // b2
+  "",                                       // b3
+
+  "",                                       // b4
+  "",                                       // b5
+  "",                                       // b6
+  ""                                        // b7
+};
+*/
+
+#ifdef _WITH_PHASE_NAME_SUPPORT
+const char * mksb_channels[] = {D_DEVICE_INPUT , D_BATTERY};
+#endif
+
+#ifdef _WITH_SERIAL_DEBUGGING
+const char * mksb_rx_counter_names[] = {
+  "Good",
+  "Timeout",
+  "TooShort",
+  "Overflow",
+  "CRC",
+  "Command",
+  "Length"
+};
+#endif
+
+// module data
+mksb_data_t me = { .Serial = nullptr, .pRxBuffer = nullptr }; // singleton
+//mksb_data_t * pMe = nullptr;
+
+
+
+uint8_t MksbChecksum(uint8_t *data, uint8_t len)
+{
+  uint8_t checksum = 0;
+  uint32_t i;
+
+  len -= 1;
+  for (i = 1; i < len; i++) { checksum += data[i]; }
+  return checksum;
+}
+
+
+unsigned long MksbExtractUint(char *data, uint8_t offset_lsb, uint8_t offset_msb)
+{
+	unsigned long result = 0;
+
+  for ( ; offset_msb >= offset_lsb; offset_msb-- ) {
+    result = (result << 8) | (uint8_t)data[offset_msb];
+  }
+	return result;
+}
+
+
+#if 0 // notused
+void MksbSetUint(unsigned long value, uint8_t *data, uint8_t offset_lsb, uint8_t offset_msb)
+{
+    for ( ; offset_lsb <= offset_msb; offset_lsb++ ) {
+      data[offset_lsb] = value & 0xFF;
+      value >>= 8;
+    }
+}
+#endif
+
+
+/********************************************************************************************/
+/* SEND DATA */
+
+bool MksbSend(uint8_t *data, uint8_t len)
+{
+  uint32_t i;
+
+  if (me.timeout) {
+//    AddLog( LOG_LEVEL_DEBUG, "NRG: Serial Busy, request pending");
+    return false; // transaction busy, a request is already waiting for response
+  }
+  me.timeout = MKSB_TIMEOUT;
+
+#ifdef _WITH_SERIAL_DEBUGGING
+  { // serial debug at LOG_LEVEL_DEBUG_MORE: Request bytes transmitted
+    char hex_char[(len * 3) + 2];
+    me.cntTx++;
+    AddLog(LOG_LEVEL_DEBUG_MORE, PSTR("NRG: Serial Tx %s"), ToHex_P(data, len, hex_char, sizeof (hex_char), ' '));
+  }
+#endif
+
+  for ( i = 0; i < len; i++ ) {
+    me.Serial->write(data[i]);
```

Comment:
Since we will send some structure in Flash on ESP8266, use `pgm_read_byte()`:

`me.Serial->write(pgm_read_byte(&data[i]));`